### DISCUSSION TITLE
Is it possible to lazily apply set traps only when first accessed?

### DISCUSSION THREAD

**Discussion by skinsapp** [pmndrs/valtio#836] [Ideas]
Currently `set` traps are applied ***eagerly*** for all nested objects recursively. For performance reasons it would be better to apply them ***lazily*** after accessed within `useSnapshot`.

This would be similar to how `useSnapshot` ***lazily*** applies `get` traps recursively to nested objects. 

So the way to build this is to have `useSnapshot` apply *both* `get` and `set` traps. And of course, the `set` trap would only be applied the first time accessed. If another component using `useSnapshot` already applied the `set` trap for the same key, obviously it wouldn't apply it again.

Has there been any research in this area or reasons why this is a bad idea? 

Is there a possibility to miss a notification from set traps before they are lazily set?

----

This rewrite would allow for use of very large state objects without worry that performance will be taxed for storing lots of rarely touched domain models. 

It seems proxy-compare would have to be more closely coupled to the code within valtio's vanilla.js file.

**Comment by dai-shi** [comment]
Thanks for opening up a suggestion. I think I understand where you come from.

We have similar discussions from different angles, but it's a design decision.
Valtio is to utilize proxy-compare and we want to separate it as it's used by some other libraries.
I think what's novel is to separate proxies for `set` and `get` traps. Separation of concerns.
I understand more integration would realize more optimization, but this is how Valtio is designed. It's also used with Vanilla JS, without using `useSnapshot` (nor proxy-compare).

That said, apart from useSnapshot, proxy-compare and `get` trap,
I think there might be some possibilities for laziness for `set` trap. I thought the eagar setup for `set` trap made sense for most use cases, but lazy setup might be useful for some use cases.
It will be a breaking change, so, it might be v3 even if it's technically possible. (v2 is already planned.)
You are free to experiment and investigate the feasibility further, but please note that anything will happen soon.

**Comment by skinsapp** [comment]
Thanks for your response. I'm not opposed to figuring it out myself. I was just looking for a hint as to whether it was not possible or problematic. Based on your response, it seems that it's doable and worth exploration. Thank you.
